torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -201,10 +200,48 @@ class ShapePropagator {
         continue;
       }
     }
-
     return tensor_types;
   }
 
+  c10::ScalarType typeToScalarType(const c10::TypePtr & type) {
+    if (auto cast = type->cast<CompleteTensorType>()) {
+      return cast->scalarType();
+    } else if(auto cast = type->cast<DimensionedTensorType>()) {
+      return cast->scalarType();
+    } else if (auto cast = type->cast<FloatType>()) {
```

Comment:
Does that work without the call to cast() though? seems it wants a typeptr which it checks via equality but iirc cast() seems to be designed to handle (potential) subtypes.

Reply:
there are other ways of checking the type than cast. Float, Bool, and Int are singleton instances. `if type == FloatType::get()`.

Reply:
In jit_type.h we have code like this:

```
  } else if (typ->isSubtypeOf(FloatType::get())) {
```

Is that also incorrect?

I think this might be to handle `OptionalType` from jit_type.h?

Reply:
`typ->isSubtypeOf(FloatType::get()))` will also work yes.

I'm not sure about optionals here. Won't the analysis be different if there is an `Optional[float]` compared to `float` ?

Reply:
ah the other issue with scalarTypeFromJitType is that it raises an error instead of returning undefined when not found.